ihaskell-display/ihaskell-widgets/src/IHaskell/Display/Widgets/Types.hs

```diff
@@ -0,0 +1,249 @@
+instance ToPairs (Attr Icon) where toPairs (Attr x) = ["icon" .= toJSON x]
+instance ToPairs (Attr ButtonStyle) where toPairs (Attr x) = ["button_style" .= toJSON x]
+instance ToPairs (Attr B64Value) where toPairs (Attr x) = ["_b64value" .= toJSON x]
+instance ToPairs (Attr ImageFormat) where toPairs (Attr x) = ["format" .= toJSON x]
+instance ToPairs (Attr BoolValue) where toPairs (Attr x) = ["value" .= toJSON x]
+
+(=::) :: sing f -> FieldType f -> Attr f
+_ =:: x = Attr x
+
+defaultWidget :: FieldType ViewName -> Rec Attr WidgetClass
+defaultWidget viewName = (SModelModule =:: "")
+                      :& (SModelName =:: "WidgetModel")
+                      :& (SViewModule =:: "")
+                      :& (SViewName =:: viewName)
+                      :& (SMsgThrottle =:: 3)
+                      :& (SVersion =:: 0)
+                      :& (SOnDisplayed =:: return ())
+                      :& RNil
+
+defaultDOMWidget :: FieldType ViewName -> Rec Attr DOMWidgetClass
+defaultDOMWidget viewName = defaultWidget viewName <+> domAttrs
+  where domAttrs = (SVisible =:: True)
+                :& (SCSS =:: [])
+                :& (SDOMClasses =:: [])
+                :& (SWidth =:: 0)
+                :& (SHeight =:: 0)
+                :& (SPadding =:: 0)
+                :& (SMargin =:: 0)
+                :& (SColor =:: "")
+                :& (SBackgroundColor =:: "")
+                :& (SBorderColor =:: "")
+                :& (SBorderWidth =:: 0)
+                :& (SBorderRadius =:: 0)
+                :& (SBorderStyle =:: DefaultBorder)
+                :& (SFontStyle =:: DefaultFont)
+                :& (SFontWeight =:: DefaultWeight)
+                :& (SFontSize =:: 0)
+                :& (SFontFamily =:: "")
+                :& RNil
+
+defaultStringWidget :: FieldType ViewName -> Rec Attr StringClass
+defaultStringWidget viewName = defaultDOMWidget viewName <+> strAttrs
+  where strAttrs = (SStringValue =:: "")
+                :& (SDisabled =:: False)
+                :& (SDescription =:: "")
+                :& (SPlaceholder =:: "")
+                :& RNil
+
+defaultBoolWidget :: FieldType ViewName -> Rec Attr BoolClass
+defaultBoolWidget viewName = defaultDOMWidget viewName <+> boolAttrs
+  where boolAttrs = (SBoolValue =:: False)
+                 :& (SDisabled =:: False)
+                 :& (SDescription =:: "")
+                 :& RNil
+
+newtype WidgetState w = WidgetState { _getState :: Rec Attr (WidgetFields w) }
+
+-- All records with ToPair instances for their Attrs will automatically have a toJSON instance now.
+instance RecAll Attr (WidgetFields w) ToPairs => ToJSON (WidgetState w) where
+  toJSON record =
+    object
+    . concat
+    . recordToList
+    . rmap (\(Compose (Dict x)) -> Const $ toPairs x)
+    $ reifyConstraint (Proxy :: Proxy ToPairs) $ _getState record
+
+data Widget (w :: WidgetType) = Widget { uuid :: UUID, state :: IORef (WidgetState w) }
```

Comment:
We already have something called `Widget`; we should avoid using the same name in different places. This creates confusion and requires import with `hiding`. So I would probably vote to rename this one rather than the old one; perhaps change this to `BuiltinWidget` or `UIWidget` really anything other than just `Widget`.